src/Verifier/SAW/Cryptol.hs

```diff
@@ -99,8 +107,42 @@ bindProp sc prop env = do
   let env' = liftEnv env
   v <- scLocalVar sc 0
   k <- scSort sc (mkSort 0)
-  return $ env' { envP = Map.insert (normalizeProp prop) (v, 0) (envP env')
-                , envS = k : envS env' }
+  return $ env' { envP = insertSupers prop [] v (envP env')
+                , envS = k : envS env'
+                }
+
+-- | When we insert a nonerasable prop into the environment, make
+--   sure to also insert all it's superclasses.  We arrange it so
+--   that every class dictionay contains the implementation of its
+--   superclass dictionaries, which can be extracted via field projections.
+insertSupers ::
+  C.Prop ->
+  [FieldName] {- Field names to project the associated class (in reverse order) -} ->
+  Term ->
+  Map C.Prop (Term, [FieldName], Int) ->
+  Map C.Prop (Term, [FieldName], Int)
+insertSupers prop fs v m
+  -- If the prop is already in the map, stop
+  | Just _ <- Map.lookup prop m = m
+
+  -- Insert the prop and check if it has any superclasses that also need to be added
+  | otherwise = Map.insert (normalizeProp prop) (v, fs, 0) $ go prop
+
+ where
+ super p f t = insertSupers (C.TCon (C.PC p) [t]) (f:fs) v
+
+ go (C.TCon (C.PC p) [t]) =
+    case p of
+      C.PRing      -> super C.PZero "ringZero" t m
+      C.PLogic     -> super C.PZero "logicZero" t m
+      C.PField     -> super C.PRing "fieldRing" t m
+      C.PIntegral  -> super C.PRing "integralRing" t m
+      C.PRound     -> super C.PField "roundField" t . super C.PCmp "roundCmp" t $ m
+      C.PCmp       -> super C.PEq "cmpEq" t m
+      C.PSignedCmp -> super C.PEq "signedCmpEq" t m
+      _ -> m
+ go _ = m
+
```

Comment:
This looks good for the current set of superclass relationships. But I wonder how well the same approach would work if we ever get any diamonds in our superclass graph in the future? I don't know if this is something we need to worry about.

Reply:
If there's diamonds in the graph, I guess you'll get a different field access path depending on exactly what order you explore the graph in.  If all our instances are confluent, it shouldn't matter, semantically.  I guess we could always prefer the access path with fewest fields, but you'd still have arbitrary choices to make for same-length paths.